source/MaterialXGenShader/ShaderStage.cpp

```diff
@@ -0,0 +1,326 @@
+{
+    auto it = _inputs.find(name);
+    if (it == _inputs.end())
+    {
+        throw ExceptionShaderGenError("No input block named '" + name + "' exists for shader stage '" + getName() + "'");
+    }
+    return *it->second;
+}
+
+const VariableBlock& ShaderStage::getInputBlock(const string& name) const
+{
+    return const_cast<ShaderStage*>(this)->getInputBlock(name);
+}
+
+VariableBlock& ShaderStage::getOutputBlock(const string& name)
+{
+    auto it = _outputs.find(name);
+    if (it == _outputs.end())
+    {
+        throw ExceptionShaderGenError("No output block named '" + name + "' exists for shader stage '" + getName() + "'");
+    }
+    return *it->second;
+}
+
+const VariableBlock& ShaderStage::getOutputBlock(const string& name) const
+{
+    return const_cast<ShaderStage*>(this)->getOutputBlock(name);
+}
+
+VariableBlock& ShaderStage::getConstantBlock()
+{
+    return _constants;
+}
+
+const VariableBlock& ShaderStage::getConstantBlock() const
+{
+    return _constants;
+}
+
+void ShaderStage::beginScope(ShaderStage::Brackets brackets)
+{
+    switch (brackets) {
+    case Brackets::BRACES:
+        beginLine();
+        _code += "{\n";
+        break;
+    case Brackets::PARENTHESES:
+        beginLine();
+        _code += "(\n";
+        break;
+    case Brackets::SQUARES:
+        beginLine();
+        _code += "[\n";
+        break;
+    case Brackets::NONE:
+        break;
+    }
+
+    ++_indentations;
+    _scopes.push(brackets);
+}
+
+void ShaderStage::endScope(bool semicolon, bool newline)
+{
+    if (_scopes.empty())
+    {
+        throw ExceptionShaderGenError("End scope called with no scope active, please check your beginScope/endScope calls");
+    }
+
+    Brackets brackets = _scopes.back();
+    _scopes.pop();
+    --_indentations;
+
+    switch (brackets) {
+    case Brackets::BRACES:
+        beginLine();
+        _code += "}";
+        break;
+    case Brackets::PARENTHESES:
+        beginLine();
+        _code += ")";
+        break;
+    case Brackets::SQUARES:
+        beginLine();
+        _code += "]";
+        break;
+    case Brackets::NONE:
+        break;
+    }
+    if (semicolon)
+        _code += ";";
+    if (newline)
+        _code += "\n";
+}
+
+void ShaderStage::beginLine()
+{
+    // TODO: Call syntax class to get indentation string
+    // Use 4 spaces as default indentation
+    static const string& INDENTATION = "    ";
+    for (int i = 0; i < _indentations; ++i)
+    {
+        _code += INDENTATION;
+    }
+}
+
+void ShaderStage::endLine(bool semicolon)
+{
+    if (semicolon)
+    {
+        _code += ";";
+    }
+    newLine();
+}
+
+void ShaderStage::newLine()
+{
+    _code += "\n";
+}
+
+void ShaderStage::addString(const string& str)
+{
+    _code += str;
+}
+
+void ShaderStage::addLine(const string& str, bool semicolon)
+{
+    beginLine();
+    addString(str);
+    endLine(semicolon);
+}
+
+void ShaderStage::addComment(const string& str)
+{
+    beginLine();
+    _code += "// " + str;
```

Comment:
Is there a syntax definition for a comment token as this could change per language ?

Reply:
Good point! I added the Syntax class as a member of Stage for this reason. But I never got around to use it. Will fix this.